### ISSUE TITLE
tmt should detect default context from the environment

### ISSUE THREAD

**Issue by stepnem** [teemtee/tmt#771]
Reproducer:
```shell
#!/bin/sh

set -e

tmt --version

cat /etc/fedora-release

testdir="$(mktemp -d)"
oldpwd="$(pwd)"

cd "$testdir"
tmt init
tmt test create -t shell adjust-test
cat <<EOF >>./adjust-test/main.fmf
adjust:
    enabled: false
    when: distro == fedora-33
EOF

tmt run -avvv provision -h local

cd "$oldpwd"
rm -rf "$testdir"
tmt clean runs -lv
```

This runs the test, even on Fedora 33, where I would expect (after having read [the documentation](https://tmt.readthedocs.io/en/latest/spec/core.html?highlight=adjust#adjust) repeatedly) the test to be skipped, same as when `enabled: false` is specified as a top level attribute in the same `main.fmf`.

Otherwise I don't understand the purpose of `adjust` used in L1 metadata [1], or the meaning of examples in the documentation (linked above) such as the following:

```yaml
# Disable a test for older distros
enabled: true
adjust:
    enabled: false
    when: distro < fedora-33
    because: the feature was added in Fedora 33
```

[1] `adjust` is [documented](https://tmt.readthedocs.io/en/latest/spec.html#metadata-specification) as L0 / core metadata, "which are common and can be used across all metadata levels".

**Comment by psss** [comment]
The `adjust` rules are evaluated based on the provided `context`. The context can be provided in this way:

    tmt --context distro=fedora-33 run

The context is correctly defined in the CI pipelines so the proper set of tests should be executed there already now. In the future we want to also auto-detect the context from the environment so that one does not have to provide it on the command line which would be very useful for local developing.

**Comment by stepnem** [comment]
Thank you for the explanation. I think the documentation (the section on `adjust` at least) ought to be more clear about this. #772 is an attempt to address that.

> In the future we want to also auto-detect the context from the environment so that one does not have to provide it on the command line which would be very useful for local developing.

Depending on what you mean by "local", `s/local/any/` (the same issue applies to VM, connect or other provisioning methods).

Also, I haven't been able to find any issue or story for this. Shall we just use this issue, perhaps adjusting the title to something like "tmt should detect default context from the environment"?

**Comment by psss** [comment]
> I think the documentation (the section on `adjust` at least) ought to be more clear about this. #772 is an attempt to address that.

Thanks for improving the docs.

> Depending on what you mean by "local"

I meant when developing tests from your laptop (using any supported provision method) so that it's as user friendly as possible.

> Also, I haven't been able to find any issue or story for this. Shall we just use this issue, perhaps adjusting the title to something like "tmt should detect default context from the environment"?

Yeah, it seems the auto-detection is not tracked yet. Feel free to open a new issue or re-use this one, whatever you prefer more.

**Comment by stepnem** [comment]
> Yeah, it seems the auto-detection is not tracked yet. Feel free to open a new issue or re-use this one, whatever you prefer more.

I've edited the summary, thank you.

**Comment by psss** [comment]
So for the start I think we could try to detect `distro` and `arch`. The interesting part of this is that the `discover` step is executed before `provision` so for example tests and plans are filtered/adjusted before we know (for sure) what the exact distro will be.

**Comment by stepnem** [comment]
> The interesting part of this is that the discover step is executed before provision so for example tests and plans are filtered/adjusted before we know (for sure) what the exact distro will be.

Indeed. Is reordering the steps a no-go? It would seem simpler (and perhaps even more correct?) than e.g. preserving the current order and bailing out later when realizing the provisioned system doesn't fit.

**Comment by pvalena** [comment]
> Indeed. Is reordering the steps a no-go? It would seem simpler (and perhaps even more correct?) than e.g. preserving the current order and bailing out later when realizing the provisioned system doesn't fit.

This could also help with arbitrary code execution in `discover` (running discover on provisioned machine, optionaly).

@psss is there any case we don't know at the time of discover what machine / method (how) / VM image / HW will be provisioned? Can `discover` somehow modify `provision` step?

_ _ _ _

_Also, as another possibility, to speed things up, we could run the provisioning in parallel, while running `discover`! So I don't find the order in any way strict, apart from cases when one of those want to "modify" the other. But those cases can be handled individually, right?_

**Comment by lukaszachy** [comment]
> is there any case we don't know at the time of discover what machine / method (how) / VM image / HW will be provisioned? Can `discover` somehow modify `provision` step?

Not as of now however multihost tests and hardware attribute might want to do that in the future.

On hacking session we briefly touched this and so far it looks tmt could have two ways of operation:
1. discover then provision (current way) where context detection is difficult / not worth implementing
2. provision then discover (new option) where context detection could be an opt-out option.

**Comment by psss** [comment]
Running `provision` before `discover` also means that sometimes we would waste time with provisioning although there are no tests to be run. See the following example:

```
> tmt run test --name smoke
/var/tmp/tmt/run-042

/plans/features/advanced
    discover
        how: fmf
        directory: /home/psss/git/tmt
        filter: tier: 3
        names: smoke
        summary: 0 tests selected
        warning: No tests found, finishing plan.

/plans/features/basic
    discover
        how: fmf
        directory: /home/psss/git/tmt
        filter: tier: 2
        names: smoke
        summary: 0 tests selected
        warning: No tests found, finishing plan.

/plans/features/core
    discover
        how: fmf
        directory: /home/psss/git/tmt
        filter: tier: 0, 1
        names: smoke
        summary: 1 test selected
    provision
        how: local
        distro: Fedora 33 (Workstation Edition)
        summary: 1 guest provisioned
    prepare
        summary: 0 preparations applied
    execute
        how: tmt
        summary: 1 test executed
    report
        how: display
        summary: 1 test passed
    finish
        summary: 0 tasks completed
```

**Comment by i386x** [comment]
> Running `provision` before `discover` also means that sometimes we would waste time with provisioning although there are no tests to be run.

What about introducing a next step after `provision`, called e.g. `adjust`, which adjust what have been discovered based on the `provision` step? The sequence of steps would then be `discover`, `provision`, `adjust`, `prepare`, `execute`, `report`, `finish`. `adjust` can be skipped if there is nothing to adjust, e.g. there are no `adjust` directives in `fmf` metadata, or the discovered context will have no effect since it is overridden from command line etc.

(`adjust` is a metastep rather then step, since in my opinion it is not necessary to introduce new CLI interface for it)

**Comment by psss** [comment]
Not sure there is a clear solution for this chicken-and-egg problem. Even with the extra `adjust` step.  Let's have a short discussion on the triage meeting.

**Comment by psss** [comment]
There seems to be a huge potential for many complications, corner cases, endless loops. E.g. detecting a context from a provisioned guest and then reloading the plan, finding out that the provision steps actually should be disabled.

As for now, not enough capacity to look into complex issues like that. Closing for now. We can revisit later.
